### DISCUSSION TITLE
FabricJS loadFromJSON()  problem / hint / advise

### DISCUSSION THREAD

**Discussion by ddreharov** [fabricjs/fabric.js#7075] [Q&A] [answered]
Hey there,

I'm searching for a solution or hint about loadFromJson() function. I'll try to explain the scenario with examples.

1. FabricJS Version - 4.3.0
2. I have a single instance of FabricJS
3. I call multiple times (example 3) `canvas.loadFromJSON()` with different JSON Configurations.

Lets say that the JSONs include background URL's  and the image size for each one is different.

1st - Background Image - 2 mb
2nd - Background Image  - 3mb 
3rd -  Background Image - 1mb (last called json)

With different size we also have different downloading and rendering times.

1st background (2mb) - 1 min
2nd background (3mb) - 2 min
3rd background (1mb) - 30 sec

**Behavior:**

Canvas loads JSON 1,2 and 3. 

While I see canvas with JSON 3 loaded I wait 30 secs to see the background image. 
In the mean time I see that in Dev Tools => Network tab => backgrounds from JSON 1 and JSON 2 are still downloading.

After 1 minute the background from JSON 1 is loaded and rendered (while I have initialized canvas with JSON 3 configuration and its objects). After 2 minutes I the background from JSON 2 is loaded and rendered on top of background form JSON 1

This is a problem because I have initialized Canvas with JSON 3 config and I expect to see only that. 
I don't expect when the resources load from previous JSON Configs to be rendered.

I expect that loadFromJSON will discard all the event listeners and callbacks for loading and rendering a configuration( if there are any ) and start a new clean load. 

Can you give me and advice or hint how can I overcome this issue. I suppose it is a bug, but I've started a discussion before going further. 

Thank you,
Dimitar

**Comment by ddreharov** [comment]
Hey @asturur , when you have time, can you have a look? Cheers

**Reply by melchiar** [reply]
@ddreharov `loadFromJSON()` doesn't discard event listeners on the canvas, it only populates the canvas with the objects in your JSON. 

http://fabricjs.com/docs/fabric.Canvas.html#loadFromJSON

**Reply by ddreharov** [reply]
@melchiar yes I've found this the bad way with the upper explanation. I'm searching for a solution. Calling `removeListeners()` doesn't work in this condition. 

Do you have any suggestions for mitigating the problem?

**Reply by asturur** [reply]
We do not have a way to , let's say, kill a loading state.
So once yo call loadFromJson, that load has to end at some point, regardless of the fact you started some other loadings in the mean time.

What you can do is avoid loading more things unless the previous loading is finished, or  work on different canvases

**Reply by ddreharov** [reply]
Unfortunately I'm not able to wait for previous loading because there is no event (or i'm not aware of) for telling me when the assets inside the canvas are loaded and also the triggering loadFromJson is not predictable. 

I'll stick with the reinitialization of the canvas since this makes more sense to me.

Thank you

**Reply by melchiar** [reply]
Why not just use the loadFromJSON callback?

**Reply by ddreharov** [reply]
@melchiar you probably don't understand the problem that I'm facing with. I'll try to explain it again.

I need to cancel the loading of the "resources" from loadFromJson when calling loadFromJson again on the same canvas instance. 

loadFromJson callback is triggered when the canvas is populated but it doesn't include the loading of the "resources" related to like background. The background could be loaded after 10 mins since calling loadFromJson (depends on the internet connection and background size - imagine a background with 20 mb of size).


Lets say we call loadFromJson and background of it will be loaded after 10 mins. While it is loading if we call again loadFromJson on the same instance, the loading of the resources from previous loadFromJson are not canceled and were applied on the second loadFromJson after they load.

I hope I've explained it better this time.

Cheers,

**Reply by asturur** [reply]
i think that what @melchiar is saying is that the json callback should actually e a clear signal that the load ended.
If you wrap it in a promise you can just wait for it.

But anyway the point is that the proper solution is mostly dictated from what are you trying to do, that isn't clear to us ( and does not have to be ).
In general reloading various jsons on the same canvas seems either a save/load/undo/redo action from a user or something that can be done on multiple canvases.

If you have json loads that take much time and the user is executing them clicking on something, and it may happen that the user clicks multiple times while one of those is still loading, then is more a UX problem rather than a fabricjs issue.

**Reply by asturur** [reply]
re-reading your answer i see you say the background is not waited for.
That shouldn't be true, and could be a bug. If you can reproduce it in a fiddle, open an issue and we can fix it.

**Reply by ddreharov** [reply]
I'll try to create a sandbox env with the issue. Thanks for the tips so far

**Reply by ddreharov** [reply]
I've created a sandbox env with the issue. Link is here: https://stackblitz.com/edit/canvas-bug-demo?file=index.ts

when you have time please check it out.

**Reply by ddreharov** [reply]
Hey there @asturur @melchiar, when you have time guys, can you have a look at the stackblitz. Thank you

**Reply by i-prikot** [reply]
Hi. Even if you cancel the event (`removeListeners`), it does not cancel the loading of images. Browser will continue loading images, see Network tab. I think you should use here Web Worker https://jsfiddle.net/i_prikot/f3uh5g18/ Call method `worker.terminate()` or abort request after loading first image or write a function that will change the src attribute of other images to empty, which should also cancel loading.

**Reply by smoke** [reply]
We have managed to implement quite some workaround here and there using RxJS and this simple function https://stackblitz.com/edit/rxjs-loadimage-kjxjbm 

Essentially we use the `loadImage` to ensure the image is in the loadable and in the Browser cache e.g.:
```
  insertImage(url: string, canvas: Canvas, imagePosition: { x: number, y: number } = { x: 25, y: 25 }) {
    loadImage(url).pipe(
      takeUntil(this.currentPageChanged$),
      tap(image => this.insertImageUrlToCanvas(image, canvas, imagePosition)),
      catchError(err => {
        apm.captureError(`Cannot load image with url: ${url}.`);
        this.logger.error(`Cannot load image with url: ${url}.`, err);
        return EMPTY;
      })
    ).subscribe();
  }
```

Here the concept in short:
 - in case the page / canvas is changed, but the image load is not complete yet we set the `src = ''` thus aborting the loading
 - in case the page / canvas is changed and the image load is complete thanks - we do no side effect
 - in case of error it is as well propagated well and can be handled

May be this concept can be used to figure out some reasonable implementation embedded in fabricjs

**Reply by ShaMan123** [reply]
I guess we can flag a loading state on canvas and terminate ongoing stuff once a new call is made.
It makes sense.
Did you try wrapping it with lodash's `_.debounce`?
In fabric 5 it's easier to workaround this limitation.
If I were you I would load from json without the image and handle it myself

**Reply by ShaMan123** [reply]
We can look into generators...
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*

**Comment by ShaMan123** [comment]
I have looked into this.
Managed to reproduce.
v5 uses promises so we can change promises to cancellable ones. And then hold a ref to the pending one and cancel it once `loadFromJSON` is called.
https://www.npmjs.com/package/p-cancelable
https://developer.mozilla.org/en-US/docs/Web/API/AbortController

**Comment by ShaMan123** [comment, marked as answer]
#7827 solves this discussion.
Could you guys test the PR? See if it works as expected
